[ethernet] Add ``interface`` configuration variable for esp-idf

rtyle opened the pull request:
# What does this implement/fix?

Adds an optional `interface` configuration variable to SPI-based ethernet types (W5500, DM9051, ENC28J60) allowing users to select the SPI host (`spi2` or `spi3`).

The variant-based SPI host selection has been moved from C++ preprocessor conditionals into Python validation, where it sets a sensible default per variant and allows the user to override it.

This is necessary when the default SPI interface collides with another component's choice. For example, the [board_m5cores3](https://github.com/m5stack/M5CoreS3-Esphome/tree/main/components/board_m5cores3) component uses SPI2_HOST on the ESP32-S3 for its display. When using a [CoreS3](https://docs.m5stack.com/en/core/CoreS3) with a [Base LAN PoE v1.2](https://docs.m5stack.com/en/base/lan_poe_v12), setting `interface: spi3` on the ethernet component resolves the conflict.

See [discussion](https://github.com/orgs/esphome/discussions/3281).

## Default SPI interface per variant

| Variant | Has SPI3 HW? | Old default | New default | Change? |
|---------|-------------|-------------|-------------|---------|
| ESP32 | yes | SPI3 | SPI3 | no |
| ESP32-S2 | yes | SPI2 | SPI2 | no |
| ESP32-S3 | yes | SPI2 | SPI2 | no |
| ESP32-C3 | no | SPI2 | SPI2 | no |
| ESP32-C5 | no | SPI2 | SPI2 | no |
| ESP32-C6 | no | SPI2 | SPI2 | no |
| ESP32-C61 | no | SPI2 | SPI2 | no |
| ESP32-H2 | no | SPI3 | SPI2 | **yes - bugfix** |
| ESP32-P4 | yes | SPI3 | SPI2 | **yes** |

ESP32-H2 was broken before — the old C++ code defaulted to SPI3 but the hardware only has SPI2. ESP32-P4 has SPI3 hardware but the old code only defaulted to SPI3 by accident (it wasn't in the `#if` variant list); users can explicitly set `interface: spi3` on P4 if needed.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5252

## Test Environment

- [ ] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
ethernet:
  type: W5500
  cs_pin: 9
  reset_pin: 7
  interrupt_pin: 14
  mosi_pin: 10
  miso_pin: 8
  clk_pin: 6
  interface: spi3
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

rtyle committed:
enhance ethernet component for spi interface choice on esp-idf

bdraco commented:
Hi! Thanks for the contribution. While I can see you've linked to a discussion, it would be really helpful to have a complete explanation of the changes directly in the PR description. 

When PR descriptions primarily link to external discussions:
- It limits the pool of potential reviewers to only those who were involved in the original discussion
- It doesn't preserve history well (discussions can be deleted, moved, or become inaccessible)
- Reviewers have to jump between multiple places to understand the context
- Future contributors looking at git history won't have the full context of why changes were made

Could you please update the PR description to include:
1. **The problem/limitation this solves** - Why is SPI interface selection needed for Ethernet components?
2. **The specific use case** - When would someone need to use `spi3` vs the default SPI interface?
3. **Technical details** - How does this integrate with the existing ESP-IDF Ethernet implementation?
4. **Any limitations or considerations** - Are there specific ESP32 variants or scenarios where this applies?

This helps reviewers who weren't part of the original discussion understand the full context and make informed decisions about the implementation. It also creates better documentation in the git history for future reference.

The example configuration is great, but understanding *why* someone would need to specify `interface: spi3` would make this much clearer for reviewers and future users.

Thanks!

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -89,6 +91,12 @@
 }
 
+SPI_INTERFACE_MAP = {
+    "spi1": "SPI1_HOST",
```
I'm not sure spi1 should be a valid option

https://espressif-docs.readthedocs-hosted.com/projects/esp-idf/en/stable/api-reference/peripherals/spi_master.html
> ESP32 integrates four SPI peripherals.

> SPI0 and SPI1 are used internally to access the ESP32’s attached flash memory and thus are currently not open to users. They share one signal bus via an arbiter.
> SPI2 and SPI3 are general purpose SPI controllers, sometimes referred to as HSPI and VSPI, respectively. They are open to users. SPI2 and SPI3 have independent signal buses with the same respective names. Each bus has three CS lines to drive up to three SPI slaves.

rtyle replied:
I don't know if it is a wise choice but it is a choice. https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/spi_master.html

rtyle replied:
Please forgive me. I am a complete newbie.

Are you saying that I should add those test*.yaml files somewhere in my PR?

bdraco replied:
No I was showing my work testing your PR. The tests should `tests/components/ethernet`

We need a test that sets the interface to `spi2` and `spi3` on idf.

bdraco replied:
> I don't know if it is a wise choice but it is a choice. [docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/spi_master.html](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/spi_master.html)

It seems like a case if YAGNI. I'd limit it to spi2/spi3 only and if there is some use case for spi1 we can wait for someone to ask to add it.

rtyle replied:
I narrowed the selection to spi2 and spi3, as requested. I also simplified the interface override logic.  I updated the PR description.

Regarding the tests/components/ethernet, the existing ones suggest some kind of test harness that includes these files in either an ard(uino) or (esp-)idf context based on their -{ard,idf}.yaml suffix. All test-* files are 1 line long and just include an associated common-* file. Should I add an interface config to all or some of the commons? Or, should I copy one of them (say, common-55500.yaml) to a test-<name-please?>.yaml and add an interface specification. do i need a file for both spi2 and spi3?

bdraco replied:
One is enough. I modified the w5500 idf test

These are "guaranteed to compile" not "guaranteed to run" tests

rtyle committed:
remove spi1 as an ethernet interface user choice on esp-idf, per espressif-docs

rtyle committed:
simplify ethernet spi interface override logic

bdraco committed:
update w5500 test to use spi3

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -218,6 +225,10 @@ def _validate(config):
                 cv.frequency, cv.int_range(int(8e6), int(80e6))
             ),
+            cv.Optional(CONF_INTERFACE): cv.All(
+                cv.only_with_framework("esp-idf"),
```
What makes this idf only?

rtyle replied:
it is based on esp-idf code.
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/spi_master.html
i modeled this "interface" configuration based on the way the spi component does the same.

rtyle replied:
it isn't anymore
https://github.com/rtyle/esphome/commit/d7362af83e55da620c0084b68eaaaf8d71ea9a74

jesserockz commented on esphome/components/ethernet/ethernet_component.cpp:
```diff
@@ -86,6 +86,11 @@ void EthernetComponent::setup() {
 #else
   auto host = SPI3_HOST;
 #endif
+#ifdef USE_ESP_IDF
+  if (this->interface_ != SPI_HOST_MAX) {
+    host = this->interface_;
+  }
+#endif
```
This feels messy, altering it after the fact. We might as well just set a default in python based on the variant which will allow cleaner code here

rtyle replied:
I purposefully tried to leave the smallest footprint possible so as not to inadvertantly trod on something I know nothing about. This very simple block of code overrides the default choice that comes immediately before it. That is the intent of this enhancement: to override the default choice by user configuration.

rtyle replied:
see commit message
https://github.com/rtyle/esphome/commit/22f38d05b79b8311a0649ee3dfaba36a32d2d2bd

jesserockz commented on esphome/components/ethernet/ethernet_component.h:
```diff
@@ -5,6 +5,14 @@
+
+#include "driver/spi_master.h"
+
+using SPIInterface = spi_host_device_t;
```
We dont really need to relabel this

rtyle replied:
This is an artifact of me modeling what was done the spi component code. They do exactly this. I can take it out if you like. Let me know.

bdraco replied:
Yes please remove it

rtyle replied:
https://github.com/rtyle/esphome/commit/d7362af83e55da620c0084b68eaaaf8d71ea9a74

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -89,6 +91,11 @@
     "GPIO17_OUT": ("CLK_OUT", 17),
 }
 
+SPI_INTERFACE_MAP = {
+    "spi2": "SPI2_HOST",
+    "spi3": "SPI3_HOST",
+}
```
This should be using the enum system properly instead of using RawExpression later in the code. The main spi code uses raw because it cannot use enum due to different platforms. Here we only have esp32 with a single enum.

```suggestion
spi_host_device_t = cg.global_ns.enum("spi_host_device_t")

SPI_INTERFACE_MAP = {
    "spi2": spi_host_device_t.SPI2_HOST,
    "spi3": spi_host_device_t.SPI3_HOST,
}
```

jesserockz replied:
We also need to validate the interface is available for the current variant

rtyle replied:
I'll try that.

> We also need to validate the interface is available for the current variant

I don't know what you mean.
I am using the choices afforded me by the esp-idf platform.

rtyle replied:
i created and used such a map in code generation. the old one needs to stay for config validation
https://github.com/rtyle/esphome/commit/5aaf15b5d1a9591eb5d8bad82f1936679de3df5b

rtyle committed:
generate code using spi_host_device_t enum

rtyle committed:
moved ethernet interface choice entirely to config

verified that default choices that were made in C++
are still made now, in config.
that is, c3, c6, s2 and s3 variants choose spi2
and everything else (h2) chooses spi3.
these default choices can be overridden by explicit config

grep -H board idf.*
idf.c3.yaml:  board: esp32-c3-devkitm-1
idf.c6.yaml:  board: esp32-c6-devkitc-1
idf.h2.yaml:  board: esp32-h2-devkitm-1
idf.s2.yaml:  board: esp32-s2-saola-1
idf.s3.yaml:  board: esp32-s3-devkitc-1

for y in idf.*; do esphome config $y 2>/dev/null | grep ' interface'; echo $y; done
  interface: spi2
idf.c3.yaml
  interface: spi2
idf.c6.yaml
  interface: spi3
idf.h2.yaml
  interface: spi2
idf.s2.yaml
  interface: spi2
idf.s3.yaml

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

rtyle committed:
remove ESP-IDF guard around ethernet interface

__init__.py
	new interface config is already guarded by

		config[CONF_TYPE] in SPI_ETHERNET_TYPES

	in (only) such cases,

	_validate
		will provide a default if one not in config
	_final_validate
		will use this to check for collision with spi component
	to_code
		will generate #define USE_ETHERNET_SPI

ethernet_component.cpp

	#ifdef USE_ETHERNET_SPI
	is the guard we want.
	we don't need (and don't want)
	#ifdef ESP-IDF

ethernet_component.h

	no need for
	using SPIInterface = spi_host_device_t;
	just use spi_host_device_t

	remove ESP_IDF guards
	they are all nested under USE_ETHERNET_SPI guards

rtyle committed:
remove ESP-IDF guard around ethernet interface

__init__.py
	new interface config is already guarded by

                    config[CONF_TYPE] in SPI_ETHERNET_TYPES

            in (only) such cases,

            _validate
                    will provide a default if one not in config
            _final_validate
                    will use this to check for collision with spi component
            to_code
                    will generate #define USE_ETHERNET_SPI

rtyle committed:
remove trailing whitespace

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

rtyle committed:
Merge branch 'dev' into enhancement/ethernet-spi-interface

bdraco committed:
Merge branch 'dev' into enhancement/ethernet-spi-interface

bdraco commented:
> I assume "you" is me. I don't know exactly what is required of me here.
> 
> I have been running my project (with this enhancement) for months.

That's fine for this case. The message is only to warn you that the memory analysis does not cover heap allocatins. If you introduced new memory allocations or stl containers than you would need to manually test.

Also there are at least 200-300 PRs waiting for reviews and only 2 full time reviewers so its normal to wait for a bit.

rtyle commented:
why is this taking so long?

rtyle committed:
Merge branch 'dev' into enhancement/ethernet-spi-interface

rtyle committed:
fix merge mistake

rtyle committed:
VARIANT_ESP32C5, C6 & C61 interface default to spi2 per pre-merge

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into enhancement/ethernet-spi-interface

# Conflicts:
#	esphome/components/ethernet/__init__.py
#	esphome/components/ethernet/ethernet_component.cpp
#	esphome/components/ethernet/ethernet_component.h

bdraco committed:
preen

bdraco committed:
address review, validate

bdraco committed:
address review, validate

bdraco committed:
address review, validate

bdraco committed:
address review, validate

bdraco committed:
address review, validate

bdraco committed:
fix validation

bdraco committed:
bot coments

rtyle committed:
Merge branch 'dev' into enhancement/ethernet-spi-interface

bdraco committed:
touch ups

bdraco committed:
touch ups

bdraco committed:
match

bdraco committed:
match

bdraco committed:
amek proper validator

bdraco committed:
preen

bdraco committed:
bot comments

bdraco committed:
Merge branch 'dev' into enhancement/ethernet-spi-interface
